wom/src/main/scala/wom/values/WomFile.scala

```diff
@@ -1,67 +1,207 @@
+sealed trait WomFile extends WomValue {
+  def value: String
 
   override def valueString = value
 
-  // TODO: WOM: WOMFILE: When WDL supports directories this check may need more work (refactoring to subclasses?)
-  override def equals(rhs: WomValue): Try[WomBoolean] = rhs match {
-    case r: WomFile => Success(WomBoolean(value.equals(r.value) && womType.equals(r.womType)))
-    case r: WomString => Success(WomBoolean(value.toString.equals(r.value.toString) && womType == WomSingleFileType))
-    case r: WomOptionalValue => evaluateIfDefined("==", r, equals)
-    case _ => invalid(s"$value == $rhs")
+  /**
+    * Converts the location using f() recursively converting any files referred to by this file.
+    *
+    * @param f The function to update the location.
+    * @return A new WomFile with the updated location.
+    */
+  def mapFile(f: String => String): WomFile
+
+  /**
+    * Returns the WomFile recursively referenced by this instance.
+    *
+    * WomSingleDirectory return either just the directory, or if there is Some listing then just the recursive listing
+    * entries. WomSingleFile return that instance plus any instances recursively discovered in secondaryFiles.
+    * WomGlobFile return just the instance.
+    */
+  def flattenFiles: Seq[WomFile] = {
+    this match {
+      case womMaybeListedDirectory: WomMaybeListedDirectory =>
+        womMaybeListedDirectory.listingOption.getOrElse(Nil).toList match {
+          case Nil => List(this)
+          case list => list.flatMap(_.flattenFiles)
+        }
+      case womMaybePopulatedFile: WomMaybePopulatedFile =>
+        womMaybePopulatedFile.secondaryFiles.foldLeft(List(this)) {
+          (womFiles, womMaybeListedDirectoryOrFile) =>
+            womFiles ++ womMaybeListedDirectoryOrFile.flattenFiles
+        }
+      case womPrimitiveFile: WomPrimitiveFile => List(womPrimitiveFile)
+    }
   }
 }
 
 object WomFile {
   def apply(fileType: WomFileType, value: String) = {
     fileType match {
+      case WomUnlistedDirectoryType => WomUnlistedDirectory(value)
       case WomSingleFileType => WomSingleFile(value)
       case WomGlobFileType => WomGlobFile(value)
+      case WomMaybeListedDirectoryType => WomMaybeListedDirectory(value)
+      case WomMaybePopulatedFileType => WomMaybePopulatedFile(value)
     }
   }
+}
 
-  def mapFile(womFile: WomFile, f: String => String): WomFile = {
-    womFile match {
-      case womSingleFile: WomSingleFile => mapSingleFile(womSingleFile, f)
-      case womGlobFile: WomGlobFile => mapGlobFile(womGlobFile, f)
-    }
+sealed trait WomPrimitiveFile extends WomFile with WomPrimitive
+
+/**
+  * A directory represented only by a path to a directory.
+  *
+  * Should not be passed into command line generation. Instead, the execution engine should create a WomListedDirectory
+  * locating the files/directories within the `value` and filling in the listing.
+  *
+  * @param value The location of the directory, possibly in the cloud.
```

Comment:
could we rename `value` to `location` or `path` or similar?

Reply:
I'm leaving it as is for this PR. It was previously called `WomFile.value` and changing it would add 113 lines of diff to sort through. Should be a very quick change for IntelliJ in a future renaming-PR.